### ISSUE TITLE
2.8.2 git shrinkwrap contains git+null

### ISSUE THREAD

**Issue by KidkArolis** [npm/npm#7961]
With 2.8.2, we've run `npm install qubitdigital/ambassador --save && npm shrinkwrap` and the shrinkwrap was updated to:

```
"ambassador": {
       "version": "1.0.0",
       "from": "qubitdigital/ambassador",
       "resolved": "git+null:git@github.com:qubitdigital/ambassador.git#a37f7f62fb1636cb0fe08fc65a233218a0fab334"
    }
```

Notice the `git+null` - that looks like it's incorrect.

**Comment by othiym23** [comment]
That is definitely incorrect. It looks like `hosted-git-info` may be returning bogus SSH URLs in some cases. I'll take a look. Thanks!

**Comment by iarna** [comment]
wow,
`git+null:git@github.com:qubitdigital/ambassador.git#a37f7f62fb1636cb0fe08fc65a233218a0fab334`
is extra invalid.

It looks like `'git+' + protocol + ':' + info.ssh()` and not 'info.sshurl()'

An actual valid URL would look like:

```
git+ssh:git@github.com/qubitdigital/ambassador.git#a37f7f62fb1636cb0fe08fc65a233218a0fab334
```

**Comment by iarna** [comment]
Ah, yup:

https://github.com/npm/npm/blob/master/lib/cache/add-remote-git.js#L87

Should be sshurl()

Which is then further mangled by:

https://github.com/npm/npm/blob/master/lib/cache/add-remote-git.js#L415-L419

To result in the confusing bit above.

**Comment by othiym23** [comment]
Thanks for tracking that down, @iarna. For cloning purposes, npm needs `.ssh()`, but for saving purposes, is needs `.sshurl()` – it may be necessary to pass the parsed `hostInfo` all the way through, at least as far as `getResolved`. I'll investigate.

**Comment by othiym23** [comment]
Fixed in 387f889c0e8fb617d9cc9a42ed0a3ec49424ab5d.  Thanks for bringing this up! This will be part of the next patch release, going out very soon (like, less than an hour).

**Comment by othiym23** [comment]
...and indeed, this was included as part of `npm@2.8.3`.

**Comment by iarna** [comment]
For the record, just using `sshurl` proved to be insufficient, as `sshurl` now returns a `git+ssh` protocol and git only actually supports the `ssh` protocol and scp-style non-urls.
